nistats/tests/test_glm_reporter.py

```diff
@@ -0,0 +1,298 @@
+                          )
+
+def test_make_contrasts_dict_with_array_of_ints():
+    test_input = np.array([1, 0, 1])
+    expected_output = {'[1 0 1]': np.array([1, 0, 1])}
+    actual_output = glmr._make_contrasts_dict(test_input)
+    assert expected_output.keys() == actual_output.keys()
+    assert np.array_equal(actual_output['[1 0 1]'],
+                          expected_output['[1 0 1]'],
+                          )
+
+def test_make_page_title_heading_with_contrast_string():
+    test_input = ('SStSSp_minus_DStDSp', None)
+    expected_output = ('Report: SStSSp_minus_DStDSp',
+                       'Statistical Report for contrasts',
+                       'SStSSp_minus_DStDSp',
+                       )
+    actual_output = glmr._make_headings(*test_input)
+    assert actual_output == expected_output
+
+
+def test_make_page_title_heading_with_contrast_list_of_strings():
+    test_input = (['SStSSp_minus_DStDSp', 'DStDSp_minus_SStSSp'],
+                  None,
+                  )
+    expected_output = ('Report: DStDSp_minus_SStSSp, SStSSp_minus_DStDSp',
+                       'Statistical Report for contrasts',
+                       'DStDSp_minus_SStSSp, SStSSp_minus_DStDSp',
+                       )
+    actual_output = glmr._make_headings(*test_input)
+    assert actual_output == expected_output
+
+
+def test_make_page_title_heading_with_contrasts_title_none():
+    test_input = ({'contrast_0': [0, 0, 1],
+                   'contrast_1': [0, 1, 1],
+                   },
+                  None,
+                  )
+    expected_output = ('Report: contrast_0, contrast_1',
+                       'Statistical Report for contrasts',
+                       'contrast_0, contrast_1',
+                       )
+    actual_output = glmr._make_headings(*test_input)
+    assert actual_output == expected_output
+
+
+def test_make_page_title_heading_with_contrasts_title_auto():
+    test_input = ({'contrast_0': [0, 0, 1],
+                   'contrast_1': [0, 1, 1],
+                   },
+                  'auto',
+                  )
+    expected_output = ('Report: contrast_0, contrast_1',
+                       'Statistical Report for contrasts',
+                       'contrast_0, contrast_1',
+                       )
+    actual_output = glmr._make_headings(*test_input)
+    assert actual_output == expected_output
+
+
+def test_make_page_title_heading_with_contrasts_title_custom():
+    test_input = ({'contrast_0': [0, 0, 1],
+                   'contrast_1': [0, 1, 1],
+                   },
+                  'Custom Title for report',
+                  )
+    expected_output = ('Custom Title for report',
+                       'Custom Title for report',
+                       '',
+                       )
+    actual_output = glmr._make_headings(*test_input)
+    assert actual_output == expected_output
+
+
+def test_make_page_title_heading_with_contrasts_none_title_custom():
+    test_input = (None,
```

Comment:
I am open to ideas.

I had originally written one function to make all the test cases and one function to run it but the former was looking ugly as sin to read, and is not very diagnosis friendly with nose (though it will be with pytest) so I rewrote each test as a separate function.

The additional tests improved the coverage and helped me catch ~4 cases that had gone unconsidered earlier. so I kept them. We have had issues filed because our tests didn't catch some cases.

The longer names are very helpful when reading our rather long logs, so I kept it this way. They are tests so it doesn't matter.

Reply:
As you know, I don't have super-strong requirements regarding syntax.
IMHO the point is that future maintainers have to figure out many tests when we need them to evolve.
But it is true that atomic tests are advantageous to detect and diagnose errors.

Reply:
I think these atomic tests are a significant improvement over the previous iteration, including being easier to understand what they are testing. There might be even better ways of writing these, but I don't have ideas about that right now. I am happy to implement whatever is better.

Reply:
I have no better suggestion atm.
Let's move on with the current tests